### ISSUE TITLE
-Werror with -Wstrict-overflow causes multiple build errors on release builds

### ISSUE THREAD

**Issue by anthonybilinski** [qTox/qTox#4699]
##### Brief Description
Commit hash:  ab12555935fe5a39f1cb3470fb2daaed5b68ffc8

Reproducible: Always

##### Steps to reproduce

1. cmake -DCMAKE_BUILD_TYPE=Release .
2. make

##### Observed Behavior
The O2 optimization causes some assumptions by the compiler that then causes warnings. I see multiple warnings like "assuming signed overflow does not occur when assuming that ..." generated from -Werror=strict-overflow, stopping the build.

Some of the errors seem extremely minor, like looping on the size of a QVector which is of signed int type.

A couple examples:
```
[  9%] Building CXX object CMakeFiles/qtox_static.dir/src/core/coreav.cpp.o
/home/abilinski/build/qTox/src/chatlog/documentcache.cpp: In member function ‘QTextDocument* DocumentCache::pop()’:
/home/abilinski/build/qTox/src/chatlog/documentcache.cpp:29:16: error: assuming signed overflow does not occur when assuming that (X - c) <= X is always true [-Werror=strict-overflow]
 QTextDocument* DocumentCache::pop()
                ^
/home/abilinski/build/qTox/src/chatlog/documentcache.cpp: In destructor ‘DocumentCache::~DocumentCache()’:
/home/abilinski/build/qTox/src/chatlog/documentcache.cpp:23:1: error: assuming signed overflow does not occur when assuming that (X - c) <= X is always true [-Werror=strict-overflow]
 DocumentCache::~DocumentCache()
 ^
cc1plus: all warnings being treated as errors
CMakeFiles/qtox_static.dir/build.make:602: recipe for target 'CMakeFiles/qtox_static.dir/src/chatlog/documentcache.cpp.o' failed
```

**Comment by sudden6** [comment]
Can you provide the full list where this causes errors so we can fix it?

**Comment by anthonybilinski** [comment]
I see some only on OBS, and different ones for different distros, I guess because of different compiler versions..

here's a union of all the warnings I see:
OBS only:
```
[  316s] /usr/src/packages/BUILD/src/persistence/settingsserializer.cpp: In member function 'void SettingsSerializer::save()':
[  316s] /usr/src/packages/BUILD/src/persistence/settingsserializer.cpp:267:9: warning: assuming signed overflow does not occur when simplifying conditional to constant [-Wstrict-overflow]
[  316s]          if (g != -1) {
[  316s]          ^~
```
local and OBS:
```
[  382s] /usr/src/packages/BUILD/src/platform/posixsignalnotifier.cpp: In function 'void detail::signalHandler(int)':
[  382s] /usr/src/packages/BUILD/src/platform/posixsignalnotifier.cpp:56:60: warning: ignoring return value of 'ssize_t write(int, const void*, size_t)', declared with attribute warn_unused_result [-Wunused-result]
[  382s]      ::write(g_signalSocketPair[0], &signum, sizeof(signum));
[  382s]                                                             ^
[  382s] /usr/src/packages/BUILD/src/platform/posixsignalnotifier.cpp: In member function 'void PosixSignalNotifier::onSignalReceived()':
[  382s] /usr/src/packages/BUILD/src/platform/posixsignalnotifier.cpp:112:67: warning: ignoring return value of 'ssize_t read(int, void*, size_t)', declared with attribute warn_unused_result [-Wunused-result]
[  382s]      ::read(detail::g_signalSocketPair[1], &signum, sizeof(signum));
[  382s]                                                                    ^
```
local only:
```
/home/abilinski/build/qTox/src/chatlog/documentcache.cpp: In member function ‘QTextDocument* DocumentCache::pop()’:
/home/abilinski/build/qTox/src/chatlog/documentcache.cpp:29:16: warning: assuming signed overflow does not occur when assuming that (X - c) <= X is always true [-Wstrict-overflow]
 QTextDocument* DocumentCache::pop()
                ^
/home/abilinski/build/qTox/src/chatlog/documentcache.cpp: In destructor ‘DocumentCache::~DocumentCache()’:
/home/abilinski/build/qTox/src/chatlog/documentcache.cpp:23:1: warning: assuming signed overflow does not occur when assuming that (X - c) <= X is always true [-Wstrict-overflow]
 DocumentCache::~DocumentCache()
 ^
```

**Comment by Diadlo** [comment]
@Talkless About the second warning. Do we want to handle unsuccess write?

**Comment by Diadlo** [comment]
The last warning seems mostly strange for me. Does anybody have idea, what lead to this behaviour?
On my machine qvector.h :
```
int size;
// ...
inline bool isEmpty() const { return d->size == 0; }
// ...
inline bool empty() const
{ return d->size == 0; }
```

**Comment by Talkless** [comment]
> @Talkless About the second warning. Do we want to handle unsuccess write?

Since we cannot even use `printf()` in signal handler (to notify about this strange occurrence), I don't know if we can do anything "useful" in case if write would fail (in case of out of memory maybe?). If `write()` fails, that would mean application would not terminate as asked, so it will be killed either by system, or by user itself anyway.

I guess all we can do is:
```
const auto result = ::write(g_signalSocketPair[0], &signum, sizeof(signum));
Q_UNUSED(result); 
```
In case of `::read()` done outside of signal handler, there could be qWarning() or similar.

**Comment by zetok** [comment]
Second error about `posixsignalnotifier.cpp` is reproducible by adding `-pedantic` to cmake args; see #4710.

**Comment by Talkless** [comment]
I get that `-Wunused-result` only after adding `-D_FORTIFY_SOURCE=1` on Debian Unstable with GCC 7.2.0.

I can't reproduce that `-Wstrict-overflow` though.

I've tried to build setting `CMAKE_CXX_FLAGS` to `-D_FORTIFY_SOURCE=2 -Wall -Wextra -Wpedantic -Werror -Wunused-result -Wstrict-overflow` and GCC gave me some unused arguments (fixable with Q_UNUSED()), some function without returned value (there is assert()), and two signed/unsigned missmatches. But not these strange warnings mentioned.

Looking at that _FORTIFY_SOURCE, it probably depends on how much source code compilation is hardened on target system by default.

**Comment by sudden6** [comment]
@Talkless maybe Debian doesn't have that `warn_unused_result` attribute set by default and enables it with _FORTIFY_SOURCE?

**Comment by sudden6** [comment]
Still not fully fixed.

**Comment by farseerfc** [comment]
I can reproduce the `-Wstrict-overflow` error on Arch Linux under `i686` arch. It will not occur under `x86_64`.

**Comment by Talkless** [comment]
@farseerfc thank's for valuable info, I'll try building on Debian i686.

**Comment by Talkless** [comment]
Yep, I've managed to reproduce that `g` warning on Debian 9 i686 VM:
```
/home/vincas/code/opensource/qTox/qTox.git/src/persistence/settingsserializer.cpp:267: error: assuming signed overflow does not occur when simplifying conditional to constant [-Werror=strict-overflow]
         if (g != -1) {
         ^~
```

I believe it's GCC compiler (optimzier) bug. It compiles fine with 64bit compiler.

One workaround could be to restrict optimizations using `volatile`:

```
if (static_cast<volatile int &>(g) != -1) {
```

I'll create GCC bug report, see what's their comment.

**Comment by Talkless** [comment]
@anthonybilinski what's the system (Distribution, architecture) where you get that `QTextDocument* DocumentCache::pop()` warning? I can't get it.

**Comment by sudden6** [comment]
Travis CI for Windows is failing because of a signed overflow error: Travis for Windows https://travis-ci.org/qTox/qTox/jobs/288788605#L2079

**Comment by Diadlo** [comment]
AFAIK, all errors was fixed. So closing. If I'm wrong, please open new issue with updated list of errors
